README.md
# LINZ NZTM2000 Tile Matrix Sets

This repository houses LINZ's official [Tile Matrix Sets (TMS)](https://www.ogc.org/standards/tms) for [NZTM2000](https://www.linz.govt.nz/data/geodetic-system/datums-projections-and-heights/projections/new-zealand-transverse-mercator-2000) ([EPGS:2193](http://epsg.io/2193)).

- [NZTM2000Quad](./raw/NZTM2000Quad.json)
- [NZTM2000](./raw/NZTM2000.json)

## NZTM2000Quad (March 2021 [Preview])

The NZTM2000 Quad TMS is a new tile matrix set to provide greater interoperability between LINZ's NZTM2000 web services (such as [LINZ's Basemaps](https://basemaps.linz.govt.nz)) and third parties.

### Key Features

- All zoom scales are from Google's [WebMercatorQuad](https://docs.opengeospatial.org/is/17-083r2/17-083r2.html#62)
- Tile matrix extent and tile extent are the same
- All zoom levels are square to work with techniques like [quad keys](https://docs.microsoft.com/en-us/bingmaps/articles/bing-maps-tile-system#tile-coordinates-and-quadkeys)
- All zoom levels increase by a factor of two
- The map is centred on [The Beehive](https://en.wikipedia.org/wiki/Beehive_(New_Zealand))
 
![NztmQuad - Zoom 1](./NztmQuadZ1.png)

*Tiles shown above were generated with [xyz-tile-debug](https://github.com/blacha/xyz-tile-debug), showing tile x,y offset, quadkey and zoom level.*

LINZ decided to create the NZTM2000Quad TMS due to compatibility issues between the original NZTM2000 TMS and third parties.

### FAQ

#### Why centre on the beehive?

No particular reason, needed a centre point and it seemed like a good idea at the time.

#### Why start at zoom level 2?

Zoom level 0 & 1 are very large and when centering around New Zealand they can overflow the bounds of NZTM2000. A number of open source tools had issues working with these large tiles.

#### Does this work in ArcGIS Online with the Eagle/Esri NZTM Basemaps?

From our intial testing this does appear to work with these other tile matrix sets.

## NZTM2000

LINZ's original tile matrix set, designed for use with cartographic tooling that needs specific carto zoom scales.

https://www.linz.govt.nz/data/linz-data-service/guides-and-documentation/nztm2000-map-tile-service-schema

## NPM Package

This repository also houses an NPM package that distributes the NZTM tile matrix sets for use in web mapping applications such as OpenLayers.

### Installation

```
yarn add @linzjs/tile-matrix-set
```

### Usage

```typescript
import {Nztm2000Quad} from '@linzjs/tile-matrix-set'

Nztm2000Quad.identifier // NZTM2000Quad
```


CHANGELOG.md
## 0.0.1 (2021-03-03)


### Features

* initial commit ([50ccf58](https://github.com/linz/NZTM2000TileMatrixSet/commit/50ccf58de99d6094e4a31abe4cb05ebe7b5e984b))





src/tms/__test__/nztm2000.quad.test.ts
import o from 'ospec';
import { Nztm2000Quad } from '../nztm2000.quad';

o.spec('NZTM2000Quad', () => {
  /** Floating points need to be within this number to be the same */
  const Eps = 0.0001;
  /** Circumference of the earth */
  const WorldCircumference = 2 * 6378137 * Math.PI;

  /**
   * Magic number to convert pixel scale to meters
   * @see http://docs.opengeospatial.org/is/17-083r2/17-083r2.html#table_2:
   */
  const PixelScaleMagic = 0.28e-3;

  const GoogleZoomScales: number[] = [];
  for (let i = 0; i < 25; i++) {
    const zoomScale = WorldCircumference / 256 / (1 << i) / PixelScaleMagic;
    GoogleZoomScales[i] = zoomScale;
  }

  o('should match google zoom scales', () => {
    for (const matrix of Nztm2000Quad.tileMatrix) {
      const existing = GoogleZoomScales.find((f) => Math.abs(matrix.scaleDenominator - f) < Eps) ?? -1;
      o(Math.abs(existing - matrix.scaleDenominator) < Eps).equals(true)(
        `Zoom:${matrix.identifier} should match google`,
      );
    }
  });

  o('should be a quad', () => {
    let expectedSize = 1;
    for (const matrix of Nztm2000Quad.tileMatrix) {
      o(matrix.matrixHeight).equals(expectedSize)(`Zoom:${matrix.identifier} should be square`);
      o(matrix.matrixWidth).equals(expectedSize)(`Zoom:${matrix.identifier} should be square`);

      expectedSize *= 2;
    }
  });

  o('should match WMTS extent', () => {
    // NZTM2000 is defined as Yx order not Xy order :(
    const topLeft = { x: Nztm2000Quad.boundingBox.lowerCorner[1], y: Nztm2000Quad.boundingBox.upperCorner[0] };
    const bottomRight = { x: Nztm2000Quad.boundingBox.upperCorner[1], y: Nztm2000Quad.boundingBox.lowerCorner[0] };

    for (const matrix of Nztm2000Quad.tileMatrix) {
      o(topLeft.x).equals(matrix.topLeftCorner[1]);
      o(topLeft.y).equals(matrix.topLeftCorner[0]);

      // Validate that the bottom right point is exactly the same as the WMTS extent
      const pixelScale = matrix.scaleDenominator * PixelScaleMagic;
      const pixelCount = matrix.matrixWidth * matrix.tileWidth;

      const matrixBottomRight = {
        x: topLeft.x + pixelCount * pixelScale,
        y: topLeft.y - pixelCount * pixelScale,
      };
      o(Math.abs(bottomRight.x - matrixBottomRight.x) < Eps).equals(true);
      o(Math.abs(bottomRight.y - matrixBottomRight.y) < Eps).equals(true);
    }
  });
});


src/tile.matrix.ts
/**
 *  Basic typing for the OSGEo TileMatrixSet JSON standard
 * @see http://schemas.opengis.net/tms/1.0/json/examples/WorldMercatorWGS84Quad.json
 */
export interface TileMatrixSetType {
  type: 'TileMatrixSetType';
  title: string;
  /** Longer description of the tile matrix set */
  abstract?: string;
  /** Unique identifer for this TileMatrixSet */
  identifier: string;

  /**
   * @example "http://www.opengis.net/def/crs/EPSG/0/3395"
   */
  supportedCRS: string;
  /**
   * Reference to well known scale set
   * @example "http://www.opengis.net/d…C/1.0/WorldMercatorWGS84"
   */
  wellKnownScaleSet?: string;
  boundingBox: {
    type: 'BoundingBoxType';
    /**
     * Reference to coordinate reference system for the bounding box
     * @example "http://www.opengis.net/def/crs/EPSG/0/3395"
     */
    crs: string;
    lowerCorner: [number, number];
    upperCorner: [number, number];
  };
  tileMatrix: TileMatrixType[];
}

export interface TileMatrixType {
  type: 'TileMatrixType';
  /**
   * Matrix identifier generally numbers as strings
   * @example "1" or "20"
   */
  identifier: string;
  scaleDenominator: number;
  topLeftCorner: [number, number];
  tileWidth: number;
  tileHeight: number;
  matrixWidth: number;
  matrixHeight: number;
}


src/tms/nztm2000.ts
import { TileMatrixSetType } from '../tile.matrix';

export const Nztm2000: TileMatrixSetType = {
  type: 'TileMatrixSetType',
  title: 'LINZ NZTM2000 Map Tile Grid',
  abstract:
    'See https://www.linz.govt.nz/data/linz-data-service/guides-and-documentation/nztm2000-map-tile-service-schema',
  identifier: 'NZTM2000',
  supportedCRS: 'https://www.opengis.net/def/crs/EPSG/0/2193',
  boundingBox: {
    type: 'BoundingBoxType',
    crs: 'https://www.opengis.net/def/crs/EPSG/0/2193',
    lowerCorner: [3087000, 274000],
    upperCorner: [7173000, 3327000],
  },
  tileMatrix: [
    {
      type: 'TileMatrixType',
      identifier: '0',
      scaleDenominator: 32000000,
      topLeftCorner: [10000000, -1000000],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 2,
      matrixHeight: 4,
    },
    {
      type: 'TileMatrixType',
      identifier: '1',
      scaleDenominator: 16000000,
      topLeftCorner: [10000000, -1000000],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 4,
      matrixHeight: 8,
    },
    {
      type: 'TileMatrixType',
      identifier: '2',
      scaleDenominator: 8000000,
      topLeftCorner: [10000000, -1000000],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 8,
      matrixHeight: 16,
    },
    {
      type: 'TileMatrixType',
      identifier: '3',
      scaleDenominator: 4000000,
      topLeftCorner: [10000000, -1000000],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 16,
      matrixHeight: 32,
    },
    {
      type: 'TileMatrixType',
      identifier: '4',
      scaleDenominator: 2000000,
      topLeftCorner: [10000000, -1000000],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 32,
      matrixHeight: 64,
    },
    {
      type: 'TileMatrixType',
      identifier: '5',
      scaleDenominator: 1000000,
      topLeftCorner: [10000000, -1000000],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 64,
      matrixHeight: 128,
    },
    {
      type: 'TileMatrixType',
      identifier: '6',
      scaleDenominator: 500000,
      topLeftCorner: [10000000, -1000000],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 128,
      matrixHeight: 256,
    },
    {
      type: 'TileMatrixType',
      identifier: '7',
      scaleDenominator: 250000,
      topLeftCorner: [10000000, -1000000],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 256,
      matrixHeight: 512,
    },
    {
      type: 'TileMatrixType',
      identifier: '8',
      scaleDenominator: 100000,
      topLeftCorner: [10000000, -1000000],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 640,
      matrixHeight: 1280,
    },
    {
      type: 'TileMatrixType',
      identifier: '9',
      scaleDenominator: 50000,
      topLeftCorner: [10000000, -1000000],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 1280,
      matrixHeight: 2560,
    },
    {
      type: 'TileMatrixType',
      identifier: '10',
      scaleDenominator: 25000,
      topLeftCorner: [10000000, -1000000],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 2560,
      matrixHeight: 5120,
    },
    {
      type: 'TileMatrixType',
      identifier: '11',
      scaleDenominator: 10000,
      topLeftCorner: [10000000, -1000000],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 6400,
      matrixHeight: 12800,
    },
    {
      type: 'TileMatrixType',
      identifier: '12',
      scaleDenominator: 5000,
      topLeftCorner: [10000000, -1000000],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 12800,
      matrixHeight: 25600,
    },
    {
      type: 'TileMatrixType',
      identifier: '13',
      scaleDenominator: 2500,
      topLeftCorner: [10000000, -1000000],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 25600,
      matrixHeight: 51200,
    },
    {
      type: 'TileMatrixType',
      identifier: '14',
      scaleDenominator: 1000,
      topLeftCorner: [10000000, -1000000],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 64000,
      matrixHeight: 128000,
    },
    {
      type: 'TileMatrixType',
      identifier: '15',
      scaleDenominator: 500,
      topLeftCorner: [10000000, -1000000],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 128000,
      matrixHeight: 256000,
    },
    {
      type: 'TileMatrixType',
      identifier: '16',
      scaleDenominator: 250,
      topLeftCorner: [10000000, -1000000],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 256000,
      matrixHeight: 512000,
    },
  ],
};


src/tms/nztm2000.quad.ts
import { TileMatrixSetType } from '../tile.matrix';

export const Nztm2000Quad: TileMatrixSetType = {
  type: 'TileMatrixSetType',
  title: 'LINZ NZTM2000Quad Map Tile Grid',
  abstract: '',
  identifier: 'NZTM2000Quad',
  supportedCRS: 'https://www.opengis.net/def/crs/EPSG/0/2193',
  boundingBox: {
    type: 'BoundingBoxType',
    crs: 'https://www.opengis.net/def/crs/EPSG/0/2193',
    lowerCorner: [419435.9938, -3260586.7284],
    upperCorner: [10438190.1652, 6758167.443],
  },
  tileMatrix: [
    {
      type: 'TileMatrixType',
      identifier: '0',
      scaleDenominator: 139770566.007179,
      topLeftCorner: [10438190.1652, -3260586.7284],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 1,
      matrixHeight: 1,
    },
    {
      type: 'TileMatrixType',
      identifier: '1',
      scaleDenominator: 69885283.0035895,
      topLeftCorner: [10438190.1652, -3260586.7284],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 2,
      matrixHeight: 2,
    },
    {
      type: 'TileMatrixType',
      identifier: '2',
      scaleDenominator: 34942641.50179475,
      topLeftCorner: [10438190.1652, -3260586.7284],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 4,
      matrixHeight: 4,
    },
    {
      type: 'TileMatrixType',
      identifier: '3',
      scaleDenominator: 17471320.750897374,
      topLeftCorner: [10438190.1652, -3260586.7284],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 8,
      matrixHeight: 8,
    },
    {
      type: 'TileMatrixType',
      identifier: '4',
      scaleDenominator: 8735660.375448687,
      topLeftCorner: [10438190.1652, -3260586.7284],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 16,
      matrixHeight: 16,
    },
    {
      type: 'TileMatrixType',
      identifier: '5',
      scaleDenominator: 4367830.1877243435,
      topLeftCorner: [10438190.1652, -3260586.7284],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 32,
      matrixHeight: 32,
    },
    {
      type: 'TileMatrixType',
      identifier: '6',
      scaleDenominator: 2183915.0938621718,
      topLeftCorner: [10438190.1652, -3260586.7284],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 64,
      matrixHeight: 64,
    },
    {
      type: 'TileMatrixType',
      identifier: '7',
      scaleDenominator: 1091957.5469310859,
      topLeftCorner: [10438190.1652, -3260586.7284],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 128,
      matrixHeight: 128,
    },
    {
      type: 'TileMatrixType',
      identifier: '8',
      scaleDenominator: 545978.7734655429,
      topLeftCorner: [10438190.1652, -3260586.7284],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 256,
      matrixHeight: 256,
    },
    {
      type: 'TileMatrixType',
      identifier: '9',
      scaleDenominator: 272989.38673277147,
      topLeftCorner: [10438190.1652, -3260586.7284],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 512,
      matrixHeight: 512,
    },
    {
      type: 'TileMatrixType',
      identifier: '10',
      scaleDenominator: 136494.69336638573,
      topLeftCorner: [10438190.1652, -3260586.7284],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 1024,
      matrixHeight: 1024,
    },
    {
      type: 'TileMatrixType',
      identifier: '11',
      scaleDenominator: 68247.34668319287,
      topLeftCorner: [10438190.1652, -3260586.7284],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 2048,
      matrixHeight: 2048,
    },
    {
      type: 'TileMatrixType',
      identifier: '12',
      scaleDenominator: 34123.67334159643,
      topLeftCorner: [10438190.1652, -3260586.7284],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 4096,
      matrixHeight: 4096,
    },
    {
      type: 'TileMatrixType',
      identifier: '13',
      scaleDenominator: 17061.836670798217,
      topLeftCorner: [10438190.1652, -3260586.7284],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 8192,
      matrixHeight: 8192,
    },
    {
      type: 'TileMatrixType',
      identifier: '14',
      scaleDenominator: 8530.918335399108,
      topLeftCorner: [10438190.1652, -3260586.7284],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 16384,
      matrixHeight: 16384,
    },
    {
      type: 'TileMatrixType',
      identifier: '15',
      scaleDenominator: 4265.459167699554,
      topLeftCorner: [10438190.1652, -3260586.7284],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 32768,
      matrixHeight: 32768,
    },
    {
      type: 'TileMatrixType',
      identifier: '16',
      scaleDenominator: 2132.729583849777,
      topLeftCorner: [10438190.1652, -3260586.7284],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 65536,
      matrixHeight: 65536,
    },
    {
      type: 'TileMatrixType',
      identifier: '17',
      scaleDenominator: 1066.3647919248886,
      topLeftCorner: [10438190.1652, -3260586.7284],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 131072,
      matrixHeight: 131072,
    },
    {
      type: 'TileMatrixType',
      identifier: '18',
      scaleDenominator: 533.1823959624443,
      topLeftCorner: [10438190.1652, -3260586.7284],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 262144,
      matrixHeight: 262144,
    },
    {
      type: 'TileMatrixType',
      identifier: '19',
      scaleDenominator: 266.59119798122214,
      topLeftCorner: [10438190.1652, -3260586.7284],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 524288,
      matrixHeight: 524288,
    },
    {
      type: 'TileMatrixType',
      identifier: '20',
      scaleDenominator: 133.29559899061107,
      topLeftCorner: [10438190.1652, -3260586.7284],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 1048576,
      matrixHeight: 1048576,
    },
    {
      type: 'TileMatrixType',
      identifier: '21',
      scaleDenominator: 66.64779949530553,
      topLeftCorner: [10438190.1652, -3260586.7284],
      tileWidth: 256,
      tileHeight: 256,
      matrixWidth: 2097152,
      matrixHeight: 2097152,
    },
  ],
};
